CMakeLists.txt

```diff
@@ -0,0 +1,164 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+cmake_minimum_required(VERSION 2.8.12)  # Ships with Ubuntu 14.04 (Trusty)
+
+project(async_grpc)
+
+set(ASYNC_GRPC_MAJOR_VERSION 0)
+set(ASYNC_GRPC_MINOR_VERSION 1)
+set(ASYNC_GRPC_PATCH_VERSION 0)
+set(ASYNC_GRPC_VERSION ${ASYNC_GRPC_MAJOR_VERSION}.${ASYNC_GRPC_MINOR_VERSION}.${ASYNC_GRPC_PATCH_VERSION})
+set(ASYNC_GRPC_SOVERSION ${ASYNC_GRPC_MAJOR_VERSION}.${ASYNC_GRPC_MINOR_VERSION})
+
+include("${PROJECT_SOURCE_DIR}/cmake/functions.cmake")
+google_initialize_async_grpc_project()
+enable_testing()
+find_package(GMock REQUIRED)
+find_package(Protobuf 3.0.0 REQUIRED)
+
+set(ALL_LIBRARY_HDRS
+    async_grpc/client.h
+    async_grpc/common/blocking_queue.h
+    async_grpc/common/make_unique.h
+    async_grpc/common/mutex.h
+    async_grpc/common/optional.h
+    async_grpc/common/port.h
+    async_grpc/common/time.h
+    async_grpc/completion_queue_thread.h
+    async_grpc/event_queue_thread.h
+    async_grpc/execution_context.h
+    async_grpc/retry.h
+    async_grpc/rpc.h
+    async_grpc/rpc_handler_interface.h
+    async_grpc/rpc_handler.h
+    async_grpc/rpc_service_method_traits.h
+    async_grpc/server.h
+    async_grpc/service.h
+    async_grpc/testing/rpc_handler_test_server.h
+    async_grpc/testing/rpc_handler_wrapper.h
+    async_grpc/type_traits.h)
+
+set(ALL_LIBRARY_SRCS
+    async_grpc/common/time.cc
+    async_grpc/completion_queue_thread.cc
+    async_grpc/event_queue_thread.cc
+    async_grpc/retry.cc
+    async_grpc/rpc.cc
+    async_grpc/server.cc
+    async_grpc/service.cc)
+
+set(ALL_TESTS
+    async_grpc/server_test.cc
+    async_grpc/type_traits_test.cc)
+
+set(INSTALL_SOURCE_HDRS ${ALL_LIBRARY_HDRS})
+
+set(ALL_PROTOS
+    async_grpc/proto/math_service.proto)
+
+set(ALL_PROTO_SRCS)
+set(ALL_PROTO_HDRS)
+foreach(RELATIVEPATH ${ALL_PROTOS})
+  get_filename_component(DIR ${RELATIVEPATH} DIRECTORY)
+  get_filename_component(FILENAME ${RELATIVEPATH} NAME_WE)
+
+  list(APPEND ALL_PROTO_SRCS "${PROJECT_BINARY_DIR}/${DIR}/${FILENAME}.pb.cc")
+  list(APPEND ALL_PROTO_HDRS "${PROJECT_BINARY_DIR}/${DIR}/${FILENAME}.pb.h")
+
+  add_custom_command(
+    OUTPUT "${PROJECT_BINARY_DIR}/${DIR}/${FILENAME}.pb.cc"
+           "${PROJECT_BINARY_DIR}/${DIR}/${FILENAME}.pb.h"
+    COMMAND  ${PROTOBUF_PROTOC_EXECUTABLE}
+    ARGS --cpp_out  ${PROJECT_BINARY_DIR} -I
+      ${PROJECT_SOURCE_DIR} ${RELATIVEPATH}
+    DEPENDS ${RELATIVEPATH}
+    COMMENT "Running C++ protocol buffer compiler on ${RELATIVEPATH}"
+    VERBATIM
+  )
+endforeach()
+set_source_files_properties(${ALL_PROTO_SRCS} ${ALL_PROTO_HDRS} PROPERTIES GENERATED TRUE)
+list(APPEND ALL_LIBRARY_HDRS ${ALL_PROTO_HDRS})
+list(APPEND ALL_LIBRARY_SRCS ${ALL_PROTO_SRCS})
+
+add_library(${PROJECT_NAME} ${ALL_LIBRARY_HDRS} ${ALL_LIBRARY_SRCS})
```

Comment:
simplification: `add_library(${PROJECT_NAME} ${ALL_LIBRARY_HDRS} ${ALL_LIBRARY_SRCS} ${ALL_PROTO_HDRS} ${ALL_PROTO_SRCS})`,
then you can drop the list append above,
replace INSTALL_SOURCE_HDRS by ALL_LIBRARY_HDRS,
drop all INSTALL_* variables.